src/rust/engine/process_execution/src/remote.rs

```diff
@@ -76,6 +80,9 @@ impl CommandRunner {
     let store = self.store.clone();
     let execute_request_result = make_execute_request(&req);
 
+    // A map that stores how much time an operation has waited last time
+    let mut backoff_per_operation: HashMap<String, u64> = HashMap::new();
```

Comment:
If we want to keep this information in a `HashMap`, it's going to need to be an `Arc<Mutex<HashMap>>` which we clone as we move into closures, so that the modifications are guarded by a `Mutex` which makes them threadsafe. The idea here is that the `Mutex` means it's safe for modification, the `Arc` means it's shareable across threads, and the `clone()` means that we can have several threads each having a reference to it.

However, I think there's something neater we can do :) Right now, the information we pass around our loop (the first argument to `future::loop_fn`, and then argument we call `future::Loop::Continue` with) is an `Operation`. If we made it a tuple: `(Operation, usize)` where the `usize` is the iteration number, that would avoid this central syncronisation point, and just calculate the expected delay based on the iteration number. (Alternatively, we could use an `(Operation, Duration)` tuple to pass around the "next wait Duration", if we wanted to just be increasing the Duration value, rather than implementing "calculate duration from iteration" logic). What do you think?

Reply:
Yeah I was trying to find a way to pass the iter number around, so thanks for the pointer! It's also nice not having to clean up the hashmap when futures are finished.

Reply:
Implemented.